### PR TITLE
Custom command system

### PR DESCRIPTION
**Pull request by defunctl**
This adds a custom command system for `so` and SquareOne projects. Custom commands are added to a project's `squareone.yml` file and then will appear as `project:command-name` if a user does a `so` inside the project's folder.

See the updated [README](https://github.com/moderntribe/square1-global-docker/blob/1323eafa1836f7f8558993d1094f3a2f294b5654/README.md#custom-commands) for the syntax and some examples.

image

### CODE REVIEW

File: app/Support/Yaml.php [hunk 1]
```diff
@@ -46,8 +45,6 @@ public function loadToConfig( $path, $configKey ): Collection {
             $loaded = array_replace_recursive( ...$loaded );
```
Comment (tarecord) [inline app/Support/Yaml.php:45]: 🌮 I know it's not a part of this, but I never knew about `array_replace_recursive()`

**Review by tarecord — APPROVED** [review]

File: app/Services/CustomCommands/CommandLoader.php [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        // Hook into the starting event and register the custom commands
+        Application::starting( function ( $artisan ) use ( $commands, $pipeline ) {
+            foreach ( $commands as $command ) {
+                // Parse the command signature
+                [ $name, $arguments, $options ] = Parser::parse( $command->signature );
+
+                $c = new ClosureCommand(
+                    $command->signature,
+                    function ( ...$parameters )
+                    use ( $command, $arguments, $options, $pipeline ) {
```
[the remark this answers is not in the document]
Comment (defunctl) [reply app/Services/CustomCommands/CommandLoader.php]: I often do split screen and the entire line was really long, but I've kept the use statement on the same line now and it does look more readable now. Have a look.

File: app/Services/CustomCommands/Runners/ServiceCommandRunner.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+        $args = $this->execArgs;
+
+        $container = $this->container->getId( $command->service );
+        $args[]    = $container;
+
+        $parameters = $command->args;
+
+        foreach ( $command->options as $name => $value ) {
+            $parameters[] = '--' . $name . '=' . $value;
```
[the remark this answers is not in the document]
Comment (defunctl) [reply app/Services/CustomCommands/Runners/ServiceCommandRunner.php]: Good question, just be being inconsistent. Updated

**Review by tarecord — APPROVED** [review]
🚀

**Comment by defunctl** [comment]
Hrmm tests all of the sudden failing for strange reasons...

**Pull request merged**